src/index/base.cpp

```diff
@@ -185,6 +178,13 @@ void BaseIndex::ThreadSync()
                            __func__, pindex->GetBlockHash().ToString());
                 return;
             }
+
+            if (last_locator_write_time + SYNC_LOCATOR_WRITE_INTERVAL < current_time) {
```

Comment:
`current_time` is stale here. Maybe instead of moving this, it'd be better to just `SetBestBlockIndex(pindex->pprev);`?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25074#discussion_r868492947

> `current_time` is stale here. Maybe instead of moving this, it'd be better to just `SetBestBlockIndex(pindex->pprev);`?

Yeah probably the reason this bug was introduced is that original author was trying to avoid `current_time` being stale. But `pindex->pprev` would be the wrong best block index to use in the `Rewind()` case above. It would be pointing to parent of the rewind block instead of the block that was actually rewound to. I guess this would be better than current behavior, but still technically incorrect, and seems like it would be fragile even if current index subclasses are ok with it.

I think using stale rewind time is ok and probably better than the alternatives, even if it seems a little wonky. This is all replaced anyway in 1819bdb7cb0634dad2e07daccfe7e8a122a943d6 from #24230, as Martin noted in https://github.com/bitcoin/bitcoin/pull/24230#discussion_r866785619

Reply:
>But pindex->pprev would be the wrong best block index to use in the Rewind() case above. It would be pointing to parent of the rewind block instead of the block that was actually rewound to. I guess this would be better than current behavior, but still technically incorrect,

I don't see how it would even be incorrect... After the rewind, `pindex` is changed to the block one beyond that which the index should be current to...?

Reply:
> I don't see how it would even be incorrect... After the rewind, `pindex` is changed to the block one beyond that which the index should be current to...?

Oops, you're right. I didn't notice NextSyncBlock was returning the next block after the fork. So I think your `pindex->pprev` suggestion is good and makes sense, and probably a little better than the current fix, though I think both fixes are fine.

Reply:
I changed it to Luke's suggestion and adjusted the commit message, thanks.
While this is simpler, `current_time` being a bit stale shouldn't ever be a serious problem in my opinion -  I think of updating the best block / committing it every 30s as a non-essential but nice-to-have feature that may save users some time in case of shutdowns, if we completely removed this block completely, it shouldn't break anything.